resources/firewall_nftables.rb

```diff
@@ -0,0 +1,163 @@
+#
+# Author:: Matthias Pausch (<m.pausch@gsi.de>)
+# Cookbook:: firewall
+# Resource:: default
+#
+# Copyright:: 2011-2019, Chef Software, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+action_class do
+  include FirewallCookbook::Helpers
+  include FirewallCookbook::Helpers::Nftables
+
+  def lookup_or_create_service(name)
+    begin
+      nftables_service = Chef.run_context.resource_collection.find(service: name)
+    rescue
+      nftables_service = service name do
+        action :nothing
+      end
+    end
+    nftables_service
+  end
+
+  def lookup_or_create_rulesfile(name)
+    begin
+      nftables_file = Chef.run_context.resource_collection.find(file: name)
+    rescue
+      nftables_file = file name do
+        action :nothing
+      end
+    end
+    nftables_file
+  end
+end
+
+unified_mode true
+
+provides :firewall, os: 'linux', platform: %w(debian) do |node|
+  node['firewall'] && node['firewall']['debian_nftables'] &&
+    node['platform_version'].to_f >= (node['platform'] == 'ubuntu' ? 20.10 : 10)
+end
+
+# for firewall implementations where ipv6 can be skipped (currently iptables-specific)
+property :ipv6_enabled, [true, false], default: true
+property :rules, Hash
+
+property :disabled, [true, false], default: false
+property :enabled, [true, false], default: true
+
+def whyrun_supported?
+  false
+end
+
+action :install do
+  return if disabled?(new_resource)
```

Comment:
I don't really understand what we're doing here. We're letting the user use a resource but if they set it as disabled, then we don't run the code.
When would the user chose to use a resource but have it disabled? I've not come across this pattern before

Reply:
The main reason this is there is probably, because I wanted to be compatible with the iptables provider. TBH, I also do not understand, why setting the resource to disabled disables all of the code completely. However, when fooling around with the firewall cookbook I found the following behaviour useful:
1. It is possible to create an invalid firewall ruleset. Generating a new ruleset before trying to restart the firewall service is important, because otherwise the cookbook fails because of a broken config and can thus never repair it.
1. Sometimes I like to read the rules before starting the firewall. Disabling the firewall but still generating the rules allows me to do that.
1. With firewalls, it can be annoying if you accidently enable it and lock yourself out. So maybe this was intended as a double safety that one is really sure to enable the firewall.

Points 1. and 2. of the above list unfortunately do not work with the current code, and it is questionable whether 3. is a valid point or not.

I can fix all this, or in fact, already have in another cookbook of mine. This resulted in much more modern and understandable code, using functions such as `edit_resource` which have been available in chef for some time.  Since the updates to the code were quite substantial: Is it okay for you if I just update this PR?

Reply:
It does seem odd to skip everything in the install. Again why use the resource then. I can see and use the disable firewall recipe to disable the service on systems in AWS as security groups are then used in place of OS firewall but maybe installed by default on image. Docker though requires firewall turned on for container networking to work.

The service though is then controlled with the service resource being added to wrapper cookbook. If you add service control to the resource I can see disabled property being used then. Not sure if that is pattern we’d want to use instead of pushing users to use the service resource along side this resource in their wrapper cookbook.

Reply:
`firewall_rule` resources search for the `firewall['default']` resource, and add their rules there. There are however two ways to do this:
1. If `firewall['default']` has not been defined yet, define it automatically and add the rule there.
2. If `firewall['default']` has not been defined yet, print a warning, and do not add the rule.

Point 1. is way more flexible, because it offers the possibility to use `firewall_rule` resources in any place one would like, while with point 2. it is mandatory that `firewall['default']` was defined somewhere before in the run list. But with point 1 it is far from obvious to a user of the cookbook that defining a `firewall_rule` resource will automagically define a `firewall['default']` resource which will put that rule to action. In this case, an attribute which controls whether the `firewall['default']` is skipped or not comes in very handy.

Since the current implementation follows point 2, it is ok to skip the check.